bokehjs/src/lib/models/tools/inspectors/hover_tool.ts

```diff
@@ -249,183 +251,149 @@ export class HoverToolView extends InspectToolView {
-              [[rx, ry], jj] = _line_hit(glyph.sxs.get(i), glyph.sys.get(i), j+1)
-              break
+            if (line_policy=="next") {
+              const [tt_sxy, jj] = _line_hit(glyph.sxs.get(i), glyph.sys.get(i), j+1)
+              return [[xs[j], ys[j]], tt_sxy, jj]
             }
-            case "nearest": {
-              [[rx, ry], jj] = _nearest_line_hit(j, geometry, sx, sy, glyph.sxs.get(i), glyph.sys.get(i))
-              data_x = glyph._xs.get(i)[jj]
-              data_y = glyph._ys.get(i)[jj]
-              break
+            if (line_policy == "nearest") {
+              const [tt_sxy, jj] = _nearest_line_hit(j, geometry, glyph.sxs.get(i), glyph.sys.get(i))
+              return [[xs[jj], ys[jj]], tt_sxy, jj]
             }
-            default:
-              throw new Error("shouldn't have happened")
-          }
+            throw new Error("shouldn't have happened")
+          })()
 
-          let index: number
-          if (renderer instanceof GlyphRenderer)
-            index = renderer.view.convert_indices_from_subset([i])[0]
-          else
-            index = i
+          const index = renderer.view.convert_indices_from_subset([i])[0]
 
           const vars = {
-            index, x, y, sx, sy, data_x, data_y,
-            segment_index: jj,
-            indices: subset_indices.multiline_indices,
+            index,
+            glyph, x, y, sx, sy, tt_x, tt_y, tt_sx, tt_sy,
             name: renderer.name,
+            indices: subset_indices.multiline_indices,
+            segment_index: jj,
           }
-          tooltips.push([rx, ry, this._render_tooltips(ds, index, vars)])
+          const rendered = this._render_tooltips(ds, vars)
+          tooltips.push([tt_sx, tt_sy, rendered])
         }
       } else {
         // handle non-multiglyphs
-        const data_x = (glyph as any)._x?.[i]
-        const data_y = (glyph as any)._y?.[i]
-
-        let rx: number
-        let ry: number
-        if (this.model.point_policy == "snap_to_data") { // and renderer.glyph.sx? and renderer.glyph.sy?
-          // Pass in our screen position so we can determine which patch we're
-          // over if there are discontinuous patches.
-          let pt = glyph.get_anchor_point(this.model.anchor, i, [sx, sy])
-          if (pt == null) {
-            pt = glyph.get_anchor_point("center", i, [sx, sy])
-            if (pt == null)
-              continue // TODO?
+        const tt_x = (glyph as any)._x?.[i]
+        const tt_y = (glyph as any)._y?.[i]
+
+        const {point_policy, anchor}  = this.model
+        const [tt_sx, tt_sy] = (function() {
+          if (point_policy == "snap_to_data") {
+            const pt = glyph.get_anchor_point(anchor, i, [sx, sy])
+            if (pt != null)
+              return [pt.x,  pt.y]
+            const ptc = glyph.get_anchor_point("center", i, [sx, sy])
+            if (ptc != null)
+              return [ptc.x,  ptc.y]
+            return [sx, sy]
```

Comment:
This is the "TODO" above, using `sx` and `sy` as in other cases.  That said, I think `glyph.get_anchor_point("center"...)` will always return a value so maybe figure out a way to express that instead.

Reply:
If that's the case, then this information has to be encoded in `get_anchor_point()`'s type, i.e.:
```ts
get_anchor_point(anchor: "center", i: number, [sx, sy]: [number, number]): {x: number, y: number}
get_anchor_point(anchor: Anchor, i: number, [sx, sy]: [number, number]): {x: number, y: number} | null

get_anchor_point(anchor: Anchor, i: number, [sx, sy]: [number, number]): {x: number, y: number} | null {
  // implementation
}

Reply:
I was too circumspect. It's not really the fact that it is center, per se. It is that, for the glyphs that happen to land on this codepath ("scatter-ish" more or less), the `get_anchor_point` always returns something for all values (including "center"). But that's not true for other kinds of glyphs, that don't take this codepath.  So, I think it's a little more complicated. I can make an issue regarding cleaning up an making anchor point reporting more consistent everywhere, if you like.